comictalker/talkerbase.py

```diff
@@ -0,0 +1,196 @@
+        2: "Network",
+        3: "Data",
+        4: "Other",
+    }
+
+    def __init__(self, source: str = "", code: int = 4, desc: str = "", sub_code: int = 0) -> None:
+        super().__init__()
+        if desc == "":
+            desc = "Unknown"
+        self.desc = desc
+        self.code = code
+        self.code_name = self.codes[code]
+        self.sub_code = sub_code
+        self.source = source
+
+    def __str__(self):
+        return f"{self.source} encountered a {self.code_name} error. {self.desc}"
+
+
+class TalkerNetworkError(TalkerError):
+    """Network class exception for information sources
+
+    Attributes:
+        sub_code -- numerical code for finer detail
+            1 -- connected refused
+            2 -- api key
+            3 -- rate limit
+            4 -- timeout
+    """
+
+    net_codes = {
+        0: "General network error.",
+        1: "The connection was refused.",
+        2: "An API key error occurred.",
+        3: "Rate limit exceeded. Please wait a bit or enter a personal key if using the default.",
+        4: "The connection timed out.",
+        5: "Number of retries exceeded.",
+    }
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.net_codes[sub_code]
+
+        super().__init__(source, 2, desc, sub_code)
+
+
+class TalkerDataError(TalkerError):
+    """Data class exception for information sources
+
+    Attributes:
+        sub_code -- numerical code for finer detail
+            1 -- unexpected data
+            2 -- malformed data
+            3 -- missing data
+    """
+
+    data_codes = {
+        0: "General data error.",
+        1: "Unexpected data encountered.",
+        2: "Malformed data encountered.",
+        3: "Missing data encountered.",
+    }
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.data_codes[sub_code]
+
+        super().__init__(source, 3, desc, sub_code)
+
+
+# Class talkers instance
+class ComicTalker:
+    """This is the class for mysource."""
+
+    def __init__(self) -> None:
+        # Identity name for the information source etc.
+        self.source_details: SourceDetails = (
+            SourceDetails()
+        )  # Can use this to test if custom talker has been configured
+        self.static_options: SourceStaticOptions = SourceStaticOptions()
+
+    def check_api_key(self, key: str, url: str):
+        raise NotImplementedError
+
+    # Search for series/volumes
+    def search_for_series(
+        self,
+        series_name: str,
+        callback: Callable[[int, int], None] | None = None,
+        refresh_cache: bool = False,
+        literal: bool = False,
+    ) -> list[ComicVolume]:
+        raise NotImplementedError
+
+    # Get issues in a series/volume
+    def fetch_issues_by_volume(self, series_id: int) -> list[ComicIssue]:
+        raise NotImplementedError
```

Comment:
These two functions return the TypedDicts defined in result types. What is actually used from it: aliases, image_url, alt_images_url, volume_id, issue_id. All of these make sense to have in the `GenericMetadata` object, is there anything else that we use that wouldn't make sense to add to the `GenericMetadata` object

Reply:
I think if the data is available then have a place for it in GM. The only question could be is: Does anything (CBL, CIX) support it? If not, then is there any use? But that could get a little unwieldy; see MU output which has `related_series`, `recommendations`, `category_recommendations` etc.

There's two descriptions too. Do an issue_desc and volume_desc?

Did you plan to add these or should I give it a go?

Reply:
No metadata anywhere, supports any of `related_series`, `recommendations`, `category_recommendations`. All of those are service related things and don't actually say anything about the content that you are retrieving. In other words if I retrieve "Fullmetal Alchemist" I don't care that it says I might like "Seven Deadly Sins" (idk if that's an actual recommendation), now that is a feature I'm sure people like on the website and if they make some sort of client for MU but it is not related to the info describing "Fullmetal Alchemist".
For that reason those will never be added into `GenericMetadata`, now if someone wants that added to `ComicIssue` or `ComicVolume`, I have no problem with that.

Reply:
I've had a better look and adding: `volume_id`, `aliases` (both volume and issue) make sense. Is `image_url` already covered by `cover_image` in GM? With `alt_images_url` I'm wonder its use but the same could be said for `image_url`. Being able to swap out the cover might be a handy feature.

Reply:
Yes, `cover_image` is the same as `image_url`.
For now though let's leave this as is we can take another look at it after it's been in use for a bit.